smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/writer/SimpleCodegenWriter.java

```diff
@@ -0,0 +1,341 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.function.BiFunction;
+import java.util.logging.Logger;
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.codegen.core.SymbolContainer;
+import software.amazon.smithy.codegen.core.SymbolDependency;
+import software.amazon.smithy.codegen.core.SymbolDependencyContainer;
+import software.amazon.smithy.codegen.core.SymbolReference;
+import software.amazon.smithy.utils.AbstractCodeWriter;
+import software.amazon.smithy.utils.SmithyUnstableApi;
+
+/**
+ * A {@code SimpleCodegenWriter} is a specialized {@link AbstractCodeWriter} that makes it
+ * easier to implement code generation that utilizes {@link Symbol}s and
+ * {@link SymbolDependency} values.
+ *
+ * <p>A {@code SimpleCodegenWriter} is expected to be subclassed, and the
+ * subclass is expected to implement language-specific functionality
+ * like writing documentation comments, tracking "imports", and adding
+ * any other kinds of helpful functionality for generating source code
+ * for a programming language.
+ *
+ * <p>The following example shows how a subclass of {@code SimpleCodegenWriter}
+ * should be created. CodegenWriters are expected to define a recursive
+ * type signature (notice that {@code MyWriter} is a generic parametric
+ * type in its own type definition).
+ *
+ * <pre>{@code
+ * public final class MyWriter extends SimpleCodegenWriter<MyWriter, MyImportContainer> {
+ *     public MyWriter(String namespace) {
+ *         super(new MyDocumentationWriter(), new MyImportContainer(namespace));
+ *     }
+ *
+ *     \@Override
+ *     public String toString() {
+ *         return getImportContainer().toString() + "\n\n" + super.toString();
+ *     }
+ *
+ *     public MyWriter someCustomMethod() {
+ *         // You can implement custom methods that are specific to whatever
+ *         // language you're implementing a generator for.
+ *         return this;
+ *     }
+ * }
+ * }</pre>
+ *
+ *
+ * <h2>Formatting symbols with "T"</h2>
+ *
+ * <p>{@code SimpleCodegenWriter} registers a default formatter for "T" that writes
+ * {@link Symbol}s and {@link SymbolReference}s. Imports needed by these types
+ * are automatically registered with {@link #addUseImports} for a {@code Symbol}
+ * and {@link #addImport} for a {@code SymbolReference}. Programming languages
+ * that have a concept of namespaces can use {@link #setRelativizeSymbols} to
+ * the namespace of the CodegenWriter, and then the default symbol formatter
+ * will relativize symbols against that namespace using {@link Symbol#relativize}
+ * when writing the symbol as a string.
+ *
+ * @param <T> The concrete type, used to provide a fluent interface.
+ * @param <U> The import container used by the writer to manage imports.
+ */
+@SmithyUnstableApi
+public class SimpleCodegenWriter<T extends SimpleCodegenWriter<T, U>, U extends ImportContainer>
```

Comment:
Maybe a better name here is `SymbolWriter` since the key thing here is that this subclass writes and tracks Symbols

Reply:
Done